### DISCUSSION TITLE
Multiple database support

### DISCUSSION THREAD

**Discussion by pcancela** [oqtane/oqtane.framework#1130] [Show and tell]
Hi everyone,

I was about to create my personal website/blog and found out about Oqtane which is a really nice slim framework that will help speed things up. I noticed however that it only supports SQL Server which in my case is not what I was hoping for. Meanwhile I had the idea to try and add MySQL support to it and I'm getting pretty happy with the result...still needs some testing and polishing here an there but I think it is ready for a demo. 

The functionality has been incorporated trying to blend in and preserve or lessen the impact on current framework's architecture. I also tried to implement it with extensibility in mind, I mean with the possibility to add other database support just by following current's implementation trails.

Please see the link below and share your opinion.

[Add MySQL database support](https://github.com/oqtane/oqtane.framework/compare/dev...pcancela:oqtane-add-mysql) 

Cheers, stay safe!

**Comment by sbwalker** [comment]
@pcancela thank you for sharing your efforts.... Cross Platform Database Support is on the roadmap for the next major release and there is a lengthy discussion on the high level goals and preferred approach here: https://github.com/oqtane/oqtane.framework/discussions/964. I will definitely take a look at your implementation to see if it can be leveraged. It would also be good to hear your thoughts on how closely you feel it aligns with #964.

**Reply by pcancela** [reply]
@sbwalker Nice... I'll definitely take a look into that #964

**Reply by sbwalker** [reply]
@pcancela I have reviewed your changes and I am impressed with the work you have done to get Oqtane to run on MySQL. The changes to the UI and transactional data access via EF Core look good. The one area where you went in a different direction than the preferred approach in #964 is related to the creation and upgrade of database entities, as you took the approach of creating dedicated SQL scripts for each database as opposed to using EF Core Migrations. For the core framework I definitely do not want to have to maintain distinct scripts for each database - I want to abstract the syntactical differences for each database and work against a single consistent API in C# code - exactly the same approach as being used for the transactional data access.

**Reply by pcancela** [reply]
Thanks @sbwalker ... it is good to know you like it. :)

I didn't have the opportunity to digest all the thoughts you guys were exchanging on #964 but the EF Core migration part stands out as one of the main aspects being discussed. Well, I have to say that I agree with you on the migration topic and I felt that having to review all sql scripts its not practical not even recommended. EF Core migrations are definitely the right tool for the job and I strongly believe the framework must embrace it. The route I took was mainly trying to blend in and follow current's framework approach, trying not to make too much of a hassle and disrupt things, it was like dancing the music that is being played. 

One thing that I felt and envision the need to refactor in a near future is the installer piece, if we keep adding logic and content for multiple database support it'll soon become difficult to maintain and manage things. An idea would be to create a 2 step wizard, first step would be where you choose the database type, and then you would load the respective razor form to handle specific configurations. The installer manager could be the orchestrator leveraging the factory pattern to handle the installation properly... food for thought.

**Reply by sbwalker** [reply]
@pcancela I totally agree that the install wizard should be broken down into components that can be dynamically loaded depending on the database type selected to allow the user to enter the required fields. These components would be used in both the install wizard as well as the Site Management - Add Site UI as it allows you to create a new tenant.

**Reply by pcancela** [reply]
@sbwalker BTW, besides knowing that the preferred approach on multiple database support is taking another direction something I completely agree and support, I always like to close chapters and take the time to get to know the solution in depth and for that I've made a last commit on my approach just to fix some missing parts namely the modules part which in order to function, all IInstallable implementations need to be injected with IEnumerable<ISqlRepository>. I picked up the blogs module project and played around with it to test the module installation piece. Here's the links if you're interested. 

- [Oqtane.Framework Fake Version](https://github.com/pcancela/oqtane.framework/tree/oqtane-fake-2.0.1)
- [Oqtane.Blogs Fake version](https://github.com/pcancela/oqtane.blogs/tree/oqtane-blog-fake-1.0.3)

**Reply by pcancela** [reply]
@sbwalker Due to the latest developments around this subject I became interested and did one more iteration over this multiple database support piece and moved away from DbUp sql scripts deployments onto EF Core migrations.

I have to say that the experience was not that bad and I managed to accomplish it. Nonetheless, I was not expecting some limitations on EF tooling (ef migrations add). Below the list of the main limitations I faced that required some creativity to be dealt with.

1. Support for multiple database providers is a difficult task, on the fly migrations out of application startup is not that easy. Here's the official documentation that highlights it very well. [Migrations with Multiple Providers](https://docs.microsoft.com/en-us/ef/core/managing-schemas/migrations/providers?tabs=dotnet-core-cli)
2. Another scenario that was not easy to overcome (db context isolation) because oqtane makes use of multiple contexts. Here's the article that helps overcome this issue. [Using a Separate Migrations Project](https://docs.microsoft.com/en-us/ef/core/managing-schemas/migrations/projects?tabs=dotnet-core-cli) 

I'm happy with the result, if you run a clean installation everything's working ok but there's still some aspects that need to be thought of, like what is the impact of this changes for external modules. I didn't test for backward compatibility yet besides thinking that is not going to be too complicated.

Please have a look and once again please share your thoughts...I believe this work can inspire the full implementation you guys are doing.

[Add EF Core migrations support](https://github.com/pcancela/oqtane.framework/tree/oqtane-add-mysql)

**Reply by hishamco** [reply]
I'm interest too to have a look to your work @pcancela 

Thanks

**Reply by cnurse** [reply]
@pcancela 

Shaun has asked me to work on the Multiple Database enhancement for 2.1, so I will download your Proof of Concept and see how it fits in with the goals Shaun and I have discussed.

Thanks

**Reply by hishamco** [reply]
It would be nice to see @cnurse in Oqtane Framework, after the great work he did in DNN and Naif Blog

**Reply by sbwalker** [reply]
@pcancela Wow! I did not expect you to continue iterating on your solution and implement Migrations - but I think it is awesome that you did! The best work is always a team effort, so I think it would be great for you and @cnurse to collaborate on this. As @cnurse mentioned above, I had shared the requirements with him for Multiple Database Support ( generally outlined in #964 ) and I think your POC will be extremely helpful.

**Reply by hishamco** [reply]
@sbwalker along with this could you please start a discussion thread outline all the features for `2.1.0` included multiple database support

**Reply by pcancela** [reply]
@cnurse I pushed another commit with some late fixes so I don't know if you have the latest. I think it is better for you to git pull the [repo](https://github.com/pcancela/oqtane.framework/tree/oqtane-add-mysql) once again just to be sure.

What you'll find: a complete and fully working implementation for new installations, didn't have the time to test for more scenarios but I believe in the end its going to be the case that it will need some tweaking here and there. I found out that the migrations management can be greatly simplified. Can talk you through about it later on.

Ping me if you have any question.

Cheers!

**Reply by cnurse** [reply]
@pancela - I have done a cursory review and I like the way you approached the multi-database aspect (I haven't yet done a deep-dive on the Migrations aspect).  One point that I think we will need to address, is that we don't want to have to add a code dependency on every database  we intend to support.

**Reply by pcancela** [reply]
@cnurse ok, ping me if you need more help.
